# Palm Tree Detection

This project was intended to detect and count palm tree image with top-down view using YOLOv8 Architecture.
I used YOLOv8 architecture because of its ability that are proven in the industry with well-documented project. YOLOv8 developed by Ulralytics also have good integration API to pytorch which can easly be deployed to a released product

## Environment Setup

I am using conda 24.11.3 with Linux Mint 21.1 system.
Before continuing the process I would like to setting up the environment with conda. Make suere you have already install the miniconda or anaconda in your machine. By following this step you will creating project environment:
1. create project folder
    
    ```bash
    mkdir palm-tree-detection-with-yolov8
    ```
2. Go to the folder
    
    ```bash
    cd mkdir palm-tree-detection-with-yolov8
    ```
3. Create conda environment in normal way

    ```bash
    conda create --name palm-tree-detection-with-yolov8 python=3.11 -y
    ```
    Due to my Linux's storage limit, I would like to make my conda env specific in my project folder. This line will create a hiden folder for the env
   
    ```bash
        conda create --prefix=.conda-palm-tree-detection-with-yolov8 python=3.11 -y
    ```

4. after finish creating, activate the environment

    ```bash
    conda activate palm-tree-detection-with-yolov8
    ```
    Or if you are using Vscode:
    <br>install this
    ```bash
    pip install ipykernel
    ```

    and select the kernel from there.
    
5. Install the Ultralytics library with CUDA
    ```bash
    conda install -c pytorch -c nvidia -c conda-forge pytorch torchvision pytorch-cuda=11.8 ultralytics
    ```
6. Move this ipynb file to the project folder


## Data Gathering

The data I used in this project was originally from https://universe.roboflow.com/cryospace-yylkk/palm-tree-label-200m-splitted and its already labled

1. In the project folder ```palm-tree-detection-with-yolov8```, create dataset folder
    ```bash
    mkdir dataset
    ```
2. Download the dataset
 
    ```bash
    !curl -L "https://universe.roboflow.com/ds/JVKquPWr8c?key=xC8w7Rxzla" > roboflow.zip;
    ```
3. Extract the data into ```dataset``` folder and make sure it looks like this:
```bash
    dataset/
    ├── train/
    │   ├── images/
    │   ├── labels/
    ├── valid/
    │   ├── images/
    │   ├── labels/
    ├── test/
    │   ├── images/
    │   ├── labels/
    └── data.yaml
```

# Librareis setup

In [1]:
#checking Python versions
!python --version

Python 3.11.11


In [2]:
from ultralytics import YOLO
import torch
import os
import cv2
import numpy as np

In [3]:
#checking the model
print(YOLO('yolov8n.pt'))

YOLO(
  (model): DetectionModel(
    (model): Sequential(
      (0): Conv(
        (conv): Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        (bn): BatchNorm2d(16, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
        (act): SiLU(inplace=True)
      )
      (1): Conv(
        (conv): Conv2d(16, 32, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        (bn): BatchNorm2d(32, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
        (act): SiLU(inplace=True)
      )
      (2): C2f(
        (cv1): Conv(
          (conv): Conv2d(32, 32, kernel_size=(1, 1), stride=(1, 1), bias=False)
          (bn): BatchNorm2d(32, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
          (act): SiLU(inplace=True)
        )
        (cv2): Conv(
          (conv): Conv2d(48, 32, kernel_size=(1, 1), stride=(1, 1), bias=False)
          (bn): BatchNorm2d(32, eps=0.001, momentum=0.03, affine=True, track_running_s

In [4]:
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0))

True
NVIDIA GeForce GTX 1650


In [5]:
!cat dataset/data.yaml

train: train/images
val: valid/images
test: test/images

names: 
  0: Palm-Tree


## Training

In [6]:
absolute_path = os.getcwd()

In [7]:
data = f'{absolute_path}/dataset/data.yaml'
data

'/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/data.yaml'

In [17]:
model = YOLO('yolov8n.pt')
# Train the model
model.train(data=data, 
            epochs=15, 
            imgsz=640, 
            batch=3,
            device=0)


New https://pypi.org/project/ultralytics/8.3.61 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.59 🚀 Python-3.11.11 torch-2.5.1 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/data.yaml, epochs=15, time=None, patience=100, batch=3, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=train18, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None

train: Scanning /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/train/labels.cache... 1902 images, 3 backgrounds, 0 corrupt: 100%|██████████| 1902/1902 [00:00<?, ?it/s]
val: Scanning /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/valid/labels.cache... 219 images, 0 backgrounds, 0 corrupt: 100%|██████████| 219/219 [00:00<?, ?it/s]


Plotting labels to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0004921875), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18
Starting training for 15 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/15      1.94G      1.711      1.331      1.447        655        640: 100%|██████████| 634/634 [01:16<00:00,  8.26it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 10.72it/s]


                   all        219      19666      0.925      0.956      0.976      0.586

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/15      1.56G       1.37     0.7424      1.219        372        640: 100%|██████████| 634/634 [01:14<00:00,  8.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.36it/s]

                   all        219      19666      0.927      0.957      0.981      0.654



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/15      1.54G      1.339     0.6787      1.197        328        640: 100%|██████████| 634/634 [01:15<00:00,  8.43it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.47it/s]

                   all        219      19666      0.925      0.955      0.979      0.634



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/15      1.59G      1.299     0.6346      1.179        427        640: 100%|██████████| 634/634 [01:17<00:00,  8.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.21it/s]

                   all        219      19666      0.932      0.966      0.984      0.705



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/15      1.48G      1.265     0.6071       1.16        489        640: 100%|██████████| 634/634 [01:16<00:00,  8.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 10.85it/s]


                   all        219      19666      0.931      0.961      0.982      0.689
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/15      1.13G      1.296     0.5947      1.217        140        640: 100%|██████████| 634/634 [01:12<00:00,  8.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00,  9.31it/s]


                   all        219      19666      0.926      0.958      0.978      0.613

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/15      1.07G      1.255     0.5678      1.186        144        640: 100%|██████████| 634/634 [01:13<00:00,  8.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.69it/s]

                   all        219      19666      0.931      0.955      0.979      0.623



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/15      1.11G      1.217     0.5529      1.178        371        640: 100%|██████████| 634/634 [01:10<00:00,  8.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.28it/s]

                   all        219      19666      0.922      0.961       0.98       0.65



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/15      1.07G      1.222     0.5486      1.181        252        640: 100%|██████████| 634/634 [01:10<00:00,  9.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 10.62it/s]

                   all        219      19666       0.93      0.958      0.982      0.656



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/15      1.01G      1.203     0.5409      1.176        264        640: 100%|██████████| 634/634 [01:11<00:00,  8.84it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.05it/s]

                   all        219      19666      0.932      0.964      0.983      0.695



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/15      1.04G      1.171     0.5219      1.163        247        640: 100%|██████████| 634/634 [01:13<00:00,  8.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00,  9.88it/s]


                   all        219      19666      0.932      0.965      0.983      0.666

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/15      1.14G      1.158     0.5106      1.142        112        640: 100%|██████████| 634/634 [01:13<00:00,  8.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.31it/s]

                   all        219      19666      0.934      0.962      0.984      0.649



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/15      1.09G      1.141     0.5019       1.14        235        640: 100%|██████████| 634/634 [01:10<00:00,  9.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.67it/s]

                   all        219      19666      0.931      0.963      0.983      0.661



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/15     0.952G      1.129     0.4955      1.133        263        640: 100%|██████████| 634/634 [01:10<00:00,  8.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 11.51it/s]

                   all        219      19666      0.932      0.962      0.984      0.618



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/15      1.12G      1.109     0.4846      1.127        206        640: 100%|██████████| 634/634 [01:11<00:00,  8.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:03<00:00, 10.69it/s]


                   all        219      19666      0.935       0.96      0.984      0.658

15 epochs completed in 0.332 hours.
Optimizer stripped from /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18/weights/last.pt, 6.2MB
Optimizer stripped from /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18/weights/best.pt, 6.2MB

Validating /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18/weights/best.pt...
Ultralytics 8.3.59 🚀 Python-3.11.11 torch-2.5.1 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
Model summary (fused): 168 layers, 3,005,843 parameters, 0 gradients


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 37/37 [00:06<00:00,  6.13it/s]


                   all        219      19666      0.932      0.967      0.984      0.705
Speed: 0.4ms preprocess, 6.9ms inference, 0.0ms loss, 1.4ms postprocess per image
Results saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train18


ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7efeeb062dd0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.048048, 

## Counting palm trees in image sample and Evaluation

In [43]:
img_source = f"{absolute_path}/dataset/palm-tree.jpeg"

In [35]:
old_model = YOLO(f'{absolute_path}/runs/detect/train16/weights/best.pt')

In [36]:
sample_palm_tree = old_model(source=img_source, save=True, save_txt=True)


image 1/1 /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/palm-tree.jpeg: 640x576 300 Palm-Trees, 7.8ms
Speed: 3.4ms preprocess, 7.8ms inference, 1.7ms postprocess per image at shape (1, 3, 640, 576)
Results saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/predict
1 label saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/predict/labels


In [27]:
sample_palm_tree = model(source=img_source, save=True, save_txt=True)


image 1/1 /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/palm-tree.jpeg: 640x576 165 Palm-Trees, 7.5ms
Speed: 19.5ms preprocess, 7.5ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 576)
Results saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train183
1 label saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train183/labels


In [37]:
# Extract detection results
boxes = sample_palm_tree[0].boxes.xyxy
confidences = sample_palm_tree[0].boxes.conf
classes = sample_palm_tree[0].boxes.cls
names = sample_palm_tree[0].names


In [44]:
palm_tree_image = cv2.imread(img_source)

In [49]:
# Define random colors for each class
colors = np.random.randint(0, 255, size=(len(names), 4), dtype="uint8")

In [58]:
#changin the box
for i, box in enumerate(boxes):
    # Extract box coordinates and convert to integers
    x_min, y_min, x_max, y_max = map(int, box)

    # Get class confidence
    class_id = int(classes[i])
    confidence = confidences[i]

    # Choose a color for the class
    # color = [int(c) for c in colors[class_id]]
    color = (252, 115, 3)
    # Draw the bounding box
    cv2.rectangle(palm_tree_image, (x_min, y_min), (x_max, y_max), color, thickness=6)

    # Create the label text
    label = f"confidence:{confidence:.2f}"

    # Get the text size for positioning
    label_size, _ = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 2)
    label_y_min = max(y_min, label_size[1] + 10)  # Ensure label doesn't go outside the image

    # Draw the label background
    cv2.rectangle(palm_tree_image, (x_min, y_min - label_size[1] - 10), (x_min + label_size[0], y_min), color, -1)

    # Put the label on the image
    cv2.putText(palm_tree_image, label, (x_min, y_min - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), thickness=1)


In [59]:
im_out_path = f'{absolute_path}/runs/detect/palm_tree_sample/new_palm_tree4.jpg'
cv2.imwrite(im_out_path, palm_tree_image)


True

### image result

<img src="/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/palm_tree_sample/new_palm_tree.jpg">

In [42]:
print(f"Palm tree detected: {len(sample_palm_tree[0].boxes)}")

Palm tree detected: 300


In [77]:
model_result = model.val()

Ultralytics 8.3.59 🚀 Python-3.11.11 torch-2.5.1 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)


val: Scanning /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/dataset/valid/labels.cache... 219 images, 0 backgrounds, 0 corrupt: 100%|██████████| 219/219 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 28/28 [00:07<00:00,  3.61it/s]


                   all        219      19666      0.936      0.963      0.984      0.699
Speed: 0.4ms preprocess, 6.8ms inference, 0.0ms loss, 2.9ms postprocess per image
Results saved to /media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train166


In [83]:
#Exporting model

model.export(format="onnx") 

Ultralytics 8.3.59 🚀 Python-3.11.11 torch-2.5.1 CPU (11th Gen Intel Core(TM) i5-11400H 2.70GHz)

PyTorch: starting from '/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train16/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 5, 8400) (6.0 MB)
requirements: Ultralytics requirements ['onnx>=1.12.0', 'onnxslim', 'onnxruntime-gpu'] not found, attempting AutoUpdate...
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.0/16.0 MB 2.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 291.5/291.5 MB 2.4 MB/s eta 0:00:0000:0100:04

requirements: AutoUpdate success ✅ 150.9s, installed 3 packages: ['onnx>=1.12.0', 'onnxslim', 'onnxruntime-gpu']
requirements: ⚠️ Restart runtime or rerun command for updates to take effect


ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: slimming with onnxslim 0.1.47...
ONNX: export success ✅ 152.3s, saved as '/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/tr

'/media/dartoyo/SSD2P2/programming/python/palm-tree-detection/runs/detect/train16/weights/best.onnx'

# Conclusion

While the model trained with only 5 epoch, it still can recognize tha palm trees, though there are still wide area containing with palm tree were not detected. As the ```sample_palm_tree``` boxes are counted, the total palm tree detected by the model is 300 and the mean Average Precision calculated across Intersection over Union (IoU) thresholds from 0.5 to 0.95 (mAP50-95) result with 69.9% of avarage precision. 



I recognize that there's always room for improvement. In future work, I would explore different YOLOv8 variants, optimize training parameters, and evaluate the model's robustness on a broader range of datasets. I am confident that with further refinement, this system can become an even more valuable tool for palm tree detection.

